base/compiler/typelimits.jl

```diff
@@ -543,39 +543,40 @@ end
 # compute typeintersect over the extended inference lattice
 # where v is in the extended lattice, and t is a Type
 function tmeet(@nospecialize(v), @nospecialize(t))
-    if isa(v, Const)
-        if !has_free_typevars(t) && !isa(v.val, t)
-            return Bottom
-        end
+    if isa(v, Type)
+        return typeintersect(v, t)
+    end
+    has_free_typevars(t) && return v
+    widev = widenconst(v)
+    if widev <: t
         return v
-    elseif isa(v, PartialStruct)
-        has_free_typevars(t) && return v
-        widev = widenconst(v)
-        if widev <: t
-            return v
-        end
-        ti = typeintersect(widev, t)
-        if ti === Bottom
-            return Bottom
-        end
+    end
+    ti = typeintersect(widev, t)
+    if ti === Bottom
+        return Bottom
+    end
+    if isa(v, PartialStruct)
         @assert widev <: Tuple
-        new_fields = Vector{Any}(undef, length(v.fields))
-        for i = 1:length(new_fields)
-            if isa(v.fields[i], Core.TypeofVararg)
-                new_fields[i] = v.fields[i]
-            else
-                new_fields[i] = tmeet(v.fields[i], widenconst(getfield_tfunc(t, Const(i))))
-                if new_fields[i] === Bottom
-                    return Bottom
-                end
+        if isa(ti, DataType) && ti.name === Tuple.name
```

Comment:
Didn't we just assert that `ti.name === Tuple.name` on the previous line?

Reply:
In a somewhat indirect way and with some assumptions on the well-behavedness of `typeinstersect`. So file this under defensive programming(and the `===` should be cheap enough that it's ok).

Reply:
But, huh, is the assertion actually valid? I think it implies that if the `PartialStruct` is for something other than a tuple, its `typ` (aka `widev`) is concrete, so that either `widev < t` or `typeintersect(widev, t) == Union{}`. While that seems to hold at the moment, it also seems to be something that could be reasonably relaxed in the future. Maybe better replace the assertion with an `if` and just return the intersection result in the `else` branch?